interfaces/builtin/power_control.go

```diff
@@ -49,6 +49,10 @@ const powerControlConnectedPlugAppArmor = `
 
 # for android kernels, see https://android.googlesource.com/kernel/msm/+/android-msm-bullhead-3.10-marshmallow-dr/Documentation/devicetree/bindings/arm/msm/lpm-levels.txt
 /sys/module/lpm_levels/parameters/sleep_disabled rw,
+
+# Needed for ACPI modules to read and set values for battery charging thresholds
+/sys/class/power_supply/BAT{,*}/charge_start_threshold rw,
```

Comment:
```
$ realpath /sys/class/power_supply/BAT0/charge_stop_threshold  
/sys/devices/LNXSYSTM:00/LNXSYBUS:00/PNP0A08:00/device:1b/PNP0C09:00/PNP0C0A:00/power_supply/BAT0/charge_stop_threshold
```

AppArmor evaluates files based on their canonical name, so I think instead you might need something like:

```
/sys/devices/**/power_supply/BAT[0-9]*/charge_stop_threshold rw,
```

You can test this by editing the existing snap profile for auto-cpufreq which should be at `/var/lib/snapd/apparmor/profiles/snap.auto-cpufreq`

And then reloading it via:

```
sudo apparmor_parser -r /var/lib/snapd/apparmor/profiles/snap.auto-cpufreq
```

Let me know what you find.

Reply:
Made and pushed suggested changes as:

`/sys/devices/**/power_supply/BAT[0-9]*/charge_stop_threshold rw,` definitely makes more sense. 

> You can test this by editing the existing snap profile for auto-cpufreq which should be at /var/lib/snapd/apparmor/profiles/snap.auto-cpufreq


for auto-cpufreq  under `/var/lib/snapd/apparmor/profiles` there are a few files:

```
snap.auto-cpufreq.auto-cpufreq
snap.auto-cpufreq.auto-cpufreq-gtk
snap.auto-cpufreq.hook.configure
snap.auto-cpufreq.service
snap-update-ns.auto-cpufreq
```

What I'm looking for exactly am I looking to test as part of these files as some of them are over 1000 lines.

Reply:
So you should add that line:

```
/sys/devices/**/power_supply/BAT[0-9]*/charge_stop_threshold rw,
```

to all of the `snap.auto-cpufreq.*` profiles, then reload them via `sudo apparmor_parser -r /var/lib/snapd/apparmor/profiles/snap.auto-cpufreq.xxx` and check that things work as you expect.

Reply:
I just rebuilt snap package locally, using the[ latest snapcraft.yaml contents](https://github.com/AdnanHodzic/auto-cpufreq/blob/master/snap/snapcraft.yaml) on `master` branch.

I installed it using `--dangerous` flag, i.e `sudo snap install auto-cpufreq_2.2.0_amd64.snap --dangerous`

I've added: 

```
/sys/devices/**/power_supply/BAT[0-9]*/charge_start_threshold rw,
/sys/devices/**/power_supply/BAT[0-9]*/charge_stop_threshold rw,
```

right before `}`  to every file that had mention of auto-cpufreq under `/var/lib/snapd/apparmor/profiles` except `snap-update-ns.auto-cpufreq`

Completely list I added these lines to:
```
snap.auto-cpufreq.auto-cpufreq
snap.auto-cpufreq.auto-cpufreq-gtk
snap.auto-cpufreq.hook.configure
snap.auto-cpufreq.service
```

Reloaded each one of the above files with `sudo apparmor_parser -r` but running auto-cpufreq still failed, i.e:

```
sudo auto-cpufreq --live

Using settings defined in /etc/auto-cpufreq.conf file

Note: You can quit live mode by pressing "ctrl+c"

----------------------------------- Warning -----------------------------------

Due to Snap package confinement limitations please consider installing auto-cpufreq using
auto-cpufreq-installer: https://github.com/AdnanHodzic/auto-cpufreq/#auto-cpufreq-installer

Unable to detect state of GNOME Power Profiles daemon service!
This daemon might interfere with auto-cpufreq and should be disabled.

Steps to perform this action using auto-cpufreq: power_helper script:
git clone https://github.com/AdnanHodzic/auto-cpufreq.git
cd auto-cpufreq/auto_cpufreq
python3 power_helper.py --gnome_power_disable

Reference: https://github.com/AdnanHodzic/auto-cpufreq#configuring-auto-cpufreq

----------------------------------- Warning -----------------------------------

Unable to detect if you are using a TLP service!
This daemon might interfere with auto-cpufreq which can lead to unexpected results.
We strongly encourage you not to use TLP unless you really know what you are doing.

-------------------------------------------------------------------------------

PermissionError(13, 'Permission denied')
Linux distro: UNKNOWN distro UNKNOWN version
Linux kernel: 6.5.0-25-generic
Processor: Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz
Cores: 8
Architecture: x86_64
Driver: intel_pstate

------------------------------ Current CPU stats ------------------------------

CPU max frequency: 4600 MHz
CPU min frequency: 400 MHz

Core    Usage   Temperature     Frequency
CPU0      0.0%       nan °C      4096 MHz
CPU1      2.0%       nan °C      4542 MHz
CPU2      1.0%       nan °C      4380 MHz
CPU3      4.0%       nan °C       400 MHz
CPU4      2.0%       nan °C      4475 MHz
CPU5      1.0%       nan °C       400 MHz
CPU6      0.0%       nan °C       400 MHz
CPU7      0.0%       nan °C      4292 MHz

---------------------------- CPU frequency scaling ----------------------------

Traceback (most recent call last):
  File "/snap/auto-cpufreq/x2/bin/auto-cpufreq", line 8, in <module>
    sys.exit(main())
  File "/snap/auto-cpufreq/x2/lib/python3.10/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/snap/auto-cpufreq/x2/lib/python3.10/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/snap/auto-cpufreq/x2/lib/python3.10/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/snap/auto-cpufreq/x2/lib/python3.10/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/snap/auto-cpufreq/x2/lib/python3.10/site-packages/auto_cpufreq/bin/auto_cpufreq.py", line 139, in main
    set_autofreq()
  File "/snap/auto-cpufreq/x2/lib/python3.10/site-packages/auto_cpufreq/core.py", line 1146, in set_autofreq
    elif charging():
  File "/snap/auto-cpufreq/x2/lib/python3.10/site-packages/auto_cpufreq/core.py", line 331, in charging
    with open(Path(power_supply_path + supply + "/type")) as f:
PermissionError: [Errno 13] Permission denied: '/sys/class/power_supply/AC/type'
```

Which seems to be the same problem that I was mentioning in [forum post](https://forum.snapcraft.io/t/request-for-kernel-module-control-interface-for-auto-cpufreq/38953/4?u=ahodzic).

What I did then is I uploaded the latest snap build to beta channel `snapcraft upload auto-cpufreq_2.2.0_amd64.snap --release beta` and re-installed auto-cpufreq from beta channel `sudo snap install auto-cpufreq --channel=beta`  and re-ran the same command:

```
sudo auto-cpufreq --live

Using settings defined in /etc/auto-cpufreq.conf file

Note: You can quit live mode by pressing "ctrl+c"
Traceback (most recent call last):
  File "/snap/auto-cpufreq/150/bin/auto-cpufreq", line 8, in <module>
    sys.exit(main())
  File "/snap/auto-cpufreq/150/lib/python3.10/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/snap/auto-cpufreq/150/lib/python3.10/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/snap/auto-cpufreq/150/lib/python3.10/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/snap/auto-cpufreq/150/lib/python3.10/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/snap/auto-cpufreq/150/lib/python3.10/site-packages/auto_cpufreq/bin/auto_cpufreq.py", line 122, in main
    battery_setup()
  File "/snap/auto-cpufreq/150/lib/python3.10/site-packages/auto_cpufreq/battery_scripts/battery.py", line 41, in battery_setup
    if lsmod("thinkpad_acpi"):
  File "/snap/auto-cpufreq/150/lib/python3.10/site-packages/auto_cpufreq/battery_scripts/battery.py", line 10, in lsmod
    output = subprocess.run(
  File "/usr/lib/python3.10/subprocess.py", line 503, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib/python3.10/subprocess.py", line 971, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.10/subprocess.py", line 1863, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
PermissionError: [Errno 13] Permission denied: 'lsmod'
```

After I moved [config file,](https://github.com/AdnanHodzic/auto-cpufreq/?tab=readme-ov-file#3-auto-cpufreq-config-file) `sudo mv /etc/auto-cpufreq.conf /tmp` and re-ran auto-cpufreq, its run looked lot more successful then before:

```
sudo auto-cpufreq --live

Note: You can quit live mode by pressing "ctrl+c"

----------------------------------- Warning -----------------------------------

Due to Snap package confinement limitations please consider installing auto-cpufreq using
auto-cpufreq-installer: https://github.com/AdnanHodzic/auto-cpufreq/#auto-cpufreq-installer

Unable to detect state of GNOME Power Profiles daemon service!
This daemon might interfere with auto-cpufreq and should be disabled.

Steps to perform this action using auto-cpufreq: power_helper script:
git clone https://github.com/AdnanHodzic/auto-cpufreq.git
cd auto-cpufreq/auto_cpufreq
python3 power_helper.py --gnome_power_disable

Reference: https://github.com/AdnanHodzic/auto-cpufreq#configuring-auto-cpufreq

----------------------------------- Warning -----------------------------------

Unable to detect if you are using a TLP service!
This daemon might interfere with auto-cpufreq which can lead to unexpected results.
We strongly encourage you not to use TLP unless you really know what you are doing.

-------------------------------------------------------------------------------

Linux distro: Ubuntu 23.10 (Mantic Minotaur)
Linux kernel: 6.5.0-25-generic
Processor: Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz
Cores: 8
Architecture: x86_64
Driver: intel_pstate

------------------------------ Current CPU stats ------------------------------

CPU max frequency: 800 MHz
CPU min frequency: 400 MHz

Core    Usage   Temperature     Frequency
CPU0     10.0%        49 °C       800 MHz
CPU1     20.4%        51 °C       800 MHz
CPU2      9.3%        49 °C       800 MHz
CPU3     10.3%        51 °C       800 MHz
CPU4     11.3%        49 °C       801 MHz
CPU5     16.5%        51 °C       800 MHz
CPU6      8.2%        49 °C       800 MHz
CPU7     10.1%        51 °C       800 MHz

CPU fan speed: 0 RPM

---------------------------- CPU frequency scaling ----------------------------

Battery is: charging

Setting to use: "performance" governor
/snap/auto-cpufreq/150/usr/bin/cpufreqctl.auto-cpufreq: line 105: echo: write error: Device or resource busy
/snap/auto-cpufreq/150/usr/bin/cpufreqctl.auto-cpufreq: line 105: echo: write error: Device or resource busy
/snap/auto-cpufreq/150/usr/bin/cpufreqctl.auto-cpufreq: line 105: echo: write error: Device or resource busy
/snap/auto-cpufreq/150/usr/bin/cpufreqctl.auto-cpufreq: line 105: echo: write error: Device or resource busy
/snap/auto-cpufreq/150/usr/bin/cpufreqctl.auto-cpufreq: line 105: echo: write error: Device or resource busy
/snap/auto-cpufreq/150/usr/bin/cpufreqctl.auto-cpufreq: line 105: echo: write error: Device or resource busy
/snap/auto-cpufreq/150/usr/bin/cpufreqctl.auto-cpufreq: line 105: echo: write error: Device or resource busy
/snap/auto-cpufreq/150/usr/bin/cpufreqctl.auto-cpufreq: line 105: echo: write error: Device or resource busy
Setting to use: "balance_performance" EPP
Setting maximum CPU frequency to 4600 Mhz

Total CPU usage: 11.8 %
Total system load: 1.83
Average temp. of all cores: 50.00 °C 

High CPU load (load average: 1.83, 1.15, 1.05)
setting turbo boost: on

-------------------------------------------------------------------------------

                "auto-cpufreq" is about to refresh ..
```

But as I said in forum post, if I install previous auto-cpufreq 2.0.0 version from (stable) release, it'll run just fine without any errors using this same config file.  As also mentioned in blog, I think the only big change that was added to config file since [v2.2.0 release ](https://github.com/AdnanHodzic/auto-cpufreq/releases/tag/v2.2.0)is ability to limit battery charging threshold.